### PR TITLE
update to Xcode 9.3/Swift 4.1

### PR DESCRIPTION
**Pull request by DivineDominion**
Pretty straightforward except the last commit: it compiled, but crashed during runtime because the `Middleware<StateType>` could not be casted to `Middleware<TestStringAppState>` (line breaks by me):

```
2018-04-05 10:00:46.912595+0200 xctest[46272:853770] Could not cast value of type
'((ReSwift.Action) -> (), () -> Swift.Optional<ReSwift.StateType>) -> ((ReSwift.Action) -> ()) -> (ReSwift.Action) -> ()' (0x106e53af8) to 
'((ReSwift.Action) -> (), () -> Swift.Optional<ReSwift_macOSTests.TestStringAppState>) -> ((ReSwift.Action) -> ()) -> (ReSwift.Action) -> ()' (0x106e53d28).
```

It suffices if one middleware is explicity cast ` as Middleware<TestStringAppState>`, but I found that rather weird-looking. It also works when written as `[a,b,c] as [Middleware<TestStringAppState>]`. But I figured it'd be a nicer practice to cast the more generic middlewares to the specific type individually. I'm open to change this, though, since it's quite noisy.

Guess that's the new ["Conditional conformance"](https://github.com/apple/swift-evolution/blob/master/proposals/0143-conditional-conformances.md) thing. Not sure what to do about it. Sounds like a bug to me.

### CODE REVIEW

**Comment by DivineDominion** [comment]
Looks like Travis only has the `xcode9.3beta` image :/

File: ReSwiftTests/StoreMiddlewareTests.swift [hunk 1]
```diff
@@ -99,7 +101,10 @@ class StoreMiddlewareTests: XCTestCase {
         let reducer = TestValueStringReducer()
         let store = Store<TestStringAppState>(reducer: reducer.handleAction,
             state: TestStringAppState(),
-            middleware: [firstMiddleware, secondMiddleware, dispatchingMiddleware])
+            middleware: [
+                firstMiddleware as Middleware<TestStringAppState>,
+                secondMiddleware as Middleware<TestStringAppState>,
+                dispatchingMiddleware as Middleware<TestStringAppState>])
```
Comment (mjarvis) [inline ReSwiftTests/StoreMiddlewareTests.swift]: I'm confused as to why the compiler is requiring this. `firstMiddleware` etc is `Middleware<StateType>`, and `TestStringAppState: StateType`.
Even doing this works:
```swift
let middleware: [Middleware<TestStringAppState>] = [
            firstMiddleware,
            secondMiddleware
        ]
        let store = Store<TestStringAppState>(reducer: reducer.handleAction,
            state: TestStringAppState(),
            middleware: middleware)
```

I think this is a swift bug. This should be compiling, as in `store.init`, `middleware: [Middleware<TestStringAppState>]`.

I'm going to verify the bug and submit it to Swift.
In the meantime, I think the pre-assignment I've provided here is a better solution than `as`

Comment (DivineDominion) [reply ReSwiftTests/StoreMiddlewareTests.swift]: Good, will change it. Can you tell me about the Swift ticket number? I'll then add a comment to it in both relevant places.

Comment (mjarvis) [reply ReSwiftTests/StoreMiddlewareTests.swift]: https://bugs.swift.org/browse/SR-7362

Comment (DivineDominion) [reply ReSwiftTests/StoreMiddlewareTests.swift]: That was quick :)

**Comment by DivineDominion** [comment]
Ready when you are :)

**Comment by mjarvis** [comment]
@DivineDominion If you wrap the changes in `#if swift(>=4.1) #else #endif` They'll be able to compile & run tests on the older xcode versions for TravisCI, then we can update it to 9.3 when they deploy the stable version.

**Comment by DivineDominion** [comment]
@mjarvis I have 0 experience with Travis; can you tell why 3/4 builds are failing, e.g. https://travis-ci.org/ReSwift/ReSwift/jobs/363238558 ?

**Comment by DivineDominion** [comment]
_Sacre coer,_ it builds! :) Thanks @mjarvis. Can you still review the other changes, or are you blocked now that you've commited something?

**Comment by ranhsd** [comment]
Hi @DivineDominion 
after I upgraded my XCODE to the latest one which uses Swift 4.1 I get a crash when I am trying to run the app 

> Could not cast value of type '((ReSwift.Action) -> (), () -> Swift.Optional<Any>) -> ((ReSwift.Action) -> ()) -> (ReSwift.Action) -> ()' (0x10505e430) to '((ReSwift.Action) -> (), () -> Swift.Optional<trench_dev.AppState>) -> ((ReSwift.Action) -> ()) -> (ReSwift.Action) -> ()' (0x10505e660).
> 2018-04-10 21:45:05.593831+0300 trench-dev[83480:10706684] Could not cast value of type '((ReSwift.Action) -> (), () -> Swift.Optional<Any>) -> ((ReSwift.Action) -> ()) -> (ReSwift.Action) -> ()' (0x10505e430) to '((ReSwift.Action) -> (), () -> Swift.Optional<trench_dev.AppState>) -> ((ReSwift.Action) -> ()) -> (ReSwift.Action) -> 

I am using cocoapods and use 

`pod 'ReSwift', :git => 'https://github.com/ReSwift/ReSwift.git'`

And I still get it. Do you know what I should do? 

Thanks.

**Comment by ranhsd** [comment]
Hi @DivineDominion  I noticed that I get this error when I use multiple middlewares in my. In case I use only one I don't get it. I really don't know how to continue ...

**Comment by mjarvis** [comment]
@ranhsd The crash is due to https://bugs.swift.org/browse/SR-7362
You can work around it by pre-defining the middleware array before you create your store:
```swift
let middleware: [Middleware<AppState>] = [..]
let store = Store(middleware: middleware)
```

**Comment by ranhsd** [comment]
@mjarvis  thanks man! You're the best!

**Comment by mjarvis** [comment]
Just to cross-reference an update -- a PR has been posted to Swift fixing the issue causing our middleware assignment issue: https://github.com/apple/swift/pull/15997

**Comment by dolanmiu** [comment]
I am experiencing this issue right now after upgrading to XCode 9.4.1

What is the temporary fix to get around this issue?

**Pull request merged**